nsot/migrations/0036_add_protocol.py

```diff
@@ -0,0 +1,33 @@
+# -*- coding: utf-8 -*-
+from __future__ import unicode_literals
+
+from django.db import migrations, models
+import django.db.models.deletion
+import django_extensions.db.fields.json
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('nsot', '0035_fix_interface_name_slug'),
+    ]
+
+    operations = [
+        migrations.CreateModel(
+            name='Protocol',
+            fields=[
+                ('id', models.AutoField(verbose_name='ID', serialize=False, auto_created=True, primary_key=True)),
+                ('_attributes_cache', django_extensions.db.fields.json.JSONField(help_text='Local cache of attributes. (Internal use only)', blank=True)),
+                ('type', models.CharField(db_index=True, max_length=8, choices=[(b'bgp', b'BGP'), (b'isis', b'IS-IS'), (b'ospf', b'OSPF')])),
+                ('asn', models.PositiveIntegerField(db_index=True)),
+                ('auth_string', models.CharField(default='', max_length=255)),
+                ('description', models.CharField(default='', max_length=255)),
+                ('circuit', models.ForeignKey(related_name='protocols', verbose_name='Circuit that this protocol is running over', to='nsot.Circuit')),
```

Comment:
I think using `circuit` here can be a bit ambiguous considering that we have a `Circuit` and `Interface` entity in NSoT. It might be better off having this split into separate fields, ie, `interface` and `circuit` and have them tie into one another if the `interface` in question is actually part of a `circuit` ?

Reply:
I guess my thoughts were around `OSPF` and `ISIS` having interfaces (logical & physical) configured into an Area/IS, rather than defining a neighbor as such.

Also, In the weird off chance that we have a physical circuit between two routers, but the peering is configured between a Loopback and a Physical Interface (dunno why anyone would do it that way), the `circuit` attribute would be misleading in that case.

Reply:
Yeah, in the case of OSPF and IS-IS I was trying to capture the local interface via the circuit (all other information in the circuit isn't needed). I agree that it's a weird indirection, but my attempt to soothe that was with the `local_interface()` method I added to `Protocol`:
https://github.com/nickpegg/nsot/blob/1722210d3d3de8698241fcf142f856cb32cb532a/nsot/models.py#L1754-L1765

From our in-person conversation, one possibility is to have both and `interface` and `circuit` fields where one of the two is required. If `circuit` is given but `interface` is not, `interface` would get filled in automatically using the local interface of the circuit. It avoids the Protocol -> Circuit -> Interface indirection, but I feel like it's a duplication of data.

This all stems from trying to cover a lot of use cases with one thing (`circuit`). Part of me feels like BGP is so different from other routing protocols that maybe it should be its own thing. By excising BGP from this model, we could also simplify this to just have `interface` instead of `circuit`. Plus there's the fact that "Protocol" is a goofy name for what is really a "Peer" or "Session" in BGP terms.

@jathanism - What are your thoughts on all this?

Reply:
I think that having `interface` or `circuit` fields would be more intuitive and simpler than enforcing that a one-sided `Circuit` be created for non-BGP `Protocol` objects.

But I think there's another consideration. We'd need to validate that `circuit` is populated when `type='bgp'`. 

It makes me feel like we might want to consider an intermediate relationship similarly how `Interface -> Assignment -> Network` relationships work to enforce these constraints. I'm not clear on how that might work yet, but I wanted to get the gears turning.